crates/wasmtime/src/component/linker.rs

```diff
@@ -230,6 +231,59 @@ impl<T> LinkerInstance<'_, T> {
         self.insert(name, Definition::Func(func.into_host_func()))
     }
 
+    /// Define a new host-provided function using dynamic types.
+    ///
+    /// `name` must refer to a function type import in `component`.  If and when
+    /// that import is invoked by the component, the specified `func` will be
+    /// called, which must return a `Val` which is an instance of the result
+    /// type of the import.
+    pub fn func_new<F: Fn(&[Val]) -> Result<Val> + Send + Sync + 'static>(
+        &mut self,
+        component: &Component,
+        name: &str,
+        func: F,
+    ) -> Result<()> {
+        self.func_new_with_context(component, name, move |_, args| func(args))
+    }
+
+    /// Define a new host-provided function using dynamic types.
+    ///
+    /// This is the same as [`LinkerInstance::func_new`] except that it accepts
+    /// a function which accepts a `StoreContextMut` as the first parameter.
+    pub fn func_new_with_context<
+        F: Fn(StoreContextMut<'_, T>, &[Val]) -> Result<Val> + Send + Sync + 'static,
+    >(
+        &mut self,
+        component: &Component,
+        name: &str,
```

Comment:
Personally I think that this needs to change in the future to taking the function type as the argument. That being said this works well enough for now and it's ok to extend this later, so I think this should stay as-is.

Otherwise though eventually this will require filling out more of the `component::types` submodule in the same manner as `wasmtime::types` which is able to describe imports/exports/etc. I think the implementation will be roughly the same as what you have already with `Arc<Types>` and `FooIndex` paired together, but again it's ok to defer this to later.